crates/video/src/video_backends/v4l2_decoder.rs

```diff
@@ -0,0 +1,845 @@
+                warn!("{}", e);
+                return Sync(video::CmdResponse::Error(InvalidParameter));
+            }
+        };
+        self.streams.insert(stream_id, stream);
+        Sync(video::CmdResponse::OkNoData)
+    }
+
+    fn destroy_stream(&mut self, stream_id: u32) -> video::CmdResponseType {
+        self.streams.remove(&stream_id);
+        Sync(video::CmdResponse::OkNoData)
+    }
+
+    fn queue_resource(
+        &mut self,
+        stream_id: u32,
+        queue_type: video::QueueType,
+        resource_id: u32,
+        timestamp: u64,
+        bytes_used: Vec<u32>,
+    ) -> video::CmdResponseType {
+        let stream = match self.streams.get_mut(&stream_id) {
+            Some(stream) => stream,
+            None => {
+                return Sync(video::CmdResponse::Error(InvalidStreamId));
+            }
+        };
+        // Clone stream to burrow the fd as inmutable
+        let fd = stream.clone().file;
+        let memory_type = stream.memory(queue_type);
+        let memory = v4l2r::memory::MemoryType::n(memory_type.to_v4l2()).unwrap();
+        let queue = v4l2r::QueueType::from_value(queue_type as u32).unwrap();
+        let bufcount = match queue_type {
+            video::QueueType::InputQueue => stream.inputq_resources.map.len(),
+            video::QueueType::OutputQueue => stream.outputq_resources.map.len(),
+        };
+        let all_created = stream.all_created(queue_type);
+        let res = match stream.find_resource_mut(resource_id, queue_type) {
+            Some(resource) => resource,
+            None => {
+                warn!("Invalid resource if: {}", resource_id);
+                return Sync(video::CmdResponse::Error(InvalidParameter));
+            }
+        };
+        if bufcount > 0 && all_created {
+            if let Err(e) = v4l2r::ioctl::reqbufs::<()>(&fd, queue, memory, bufcount as u32) {
+                warn!("reqbufs failed: {}", e);
+                return Sync(video::CmdResponse::Error(InvalidParameter));
+            }
+            res.set_queried();
+        }
+
+        if res.planes.len() != bytes_used.len() {
+            warn!("Number of planes mismatch");
+            return Sync(video::CmdResponse::Error(InvalidParameter));
+        }
+        res.buffer_data.write().unwrap().timestamp = timestamp;
+        // Assumes UserPtrHandle
+        if memory_type == video::MemoryType::VirtioObject {
+            warn!("Virtio Object Memory NOT YET supported");
+            return Sync(video::CmdResponse::Error(InvalidParameter));
+        }
+        let planes: Vec<v4l2r::ioctl::QBufPlane> = res
+            .planes
+            .iter()
+            .map(|plane| {
+                v4l2r::ioctl::QBufPlane(v4l2r::bindings::v4l2_plane {
+                    length: plane.length,
+                    m: v4l2r::bindings::v4l2_plane__bindgen_ty_1 {
+                        userptr: plane.address,
+                    },
+                    ..unsafe { mem::zeroed() }
```

Comment:
Could this maybe be moved to some `Default` implementation on the struct? Then we drop the unsafe here.

Reply:
Unfortunately, this is also a burrowed struct from the library. It has a couple of constructors, but I am not sure if I can use it to point to the memory region that I want. I tried already IIRC and this was my best try.
```rust
m: v4l2r::bindings::v4l2_plane__bindgen_ty_1 {
    userptr: plane.address,
},
```
This makes the host buffer to point to the shared memory region, so that the guest can pick up the contents afterward. I'll try again to find a better solution.

Reply:
Would sending a PR upstream to add the missing Default impls / constructors make sense here?

Reply:
I've changed this part and used one of the intended constructors from the library instead of zeroing here.